src/main/java/com/hubspot/jinjava/util/ChunkResolver.java

```diff
@@ -0,0 +1,355 @@
+package com.hubspot.jinjava.util;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.hubspot.jinjava.interpret.DeferredValueException;
+import com.hubspot.jinjava.interpret.JinjavaInterpreter;
+import com.hubspot.jinjava.interpret.UnknownTokenException;
+import com.hubspot.jinjava.objects.date.JsonPyishDateSerializer;
+import com.hubspot.jinjava.objects.date.PyishDate;
+import com.hubspot.jinjava.tree.parse.Token;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * This class takes a string and resolves it in chunks. This allows for
+ * strings with deferred values within them to be partially resolved, as much
+ * as they can be with a deferred value.
+ * E.g with foo=3, bar=2:
+ *   "range(0,foo)[-1] + deferred/bar" -> "2 + deferred/2"
+ * This class is not thread-safe. Do not reuse between threads.
+ */
+public class ChunkResolver {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper()
+  .registerModule(
+      new SimpleModule().addSerializer(PyishDate.class, new JsonPyishDateSerializer())
+    );
+
+  private static final Set<String> RESERVED_KEYWORDS = ImmutableSet.of(
+    "and",
```

Comment:
can you reference constants from the tags themselves here?

Reply:
Got a list of reserved keywords from here https://svn.python.org/projects/external/Jinja-1.1/docs/build/designerdoc.html#reserved-keywords though I think it makes sense to not hardcode as many of these as possible. Just don't know where to get a more complete list, but I'll use the constants from tags to remove some of these

Reply:
Thinking about this some more, this doesn't need to include any of the tag names as we only need to look for words that appear within expressions like `in` and `or`, etc.